### PR TITLE
Terrascan init and config handling refactor

### LINKED ISSUE #550: terrascan init command doesn't work with -c flag
* terrascan version: v1.3.2
* Operating System: MacOS 10.15.7

### Description

terrascan init doesn't produce desired results with config file passed using -c flag

### What I Did

```
`terrascan init -c _pathToConfigFile_`
```

### LINKED ISSUE #570: terrascan scan command doesn't work with TERRASCAN_CONFIG env variable
* terrascan version: v1.3.3
* Operating System: MacOS 10.15.7

### Description

terrascan scan command doesn't produce desired results when config file is specified using TERRASCAN_CONFIG env variable

### What I Did

TERRASCAN_CONFIG=_pathToConfigFile_ terrascan scan -d pathToIacDir

The config file had scan and skip rules, which were not honoured when scan results were displayed.

### LINKED ISSUE #619: terrascan scan fails if path and rego_subdir are not provided together in the toml configfile
* terrascan version: master
* Operating System: mac os

### What I Did
 
```
╭─dev-gaur@Devangs-MacBook-Pro ~/go/src/github.com/accurics/terrascan ‹master*›
╰─$ rm -rf ~/.terrascan/                                                                                                                                                                                                                                  3 ↵
╭─dev-gaur@Devangs-MacBook-Pro ~/go/src/github.com/accurics/terrascan ‹master*›
╰─$ TERRASCAN_CONFIG=demo/config.toml ./bin/terrascan scan -d pkg/iac-providers/terraform/v12/testdata/deep-modules/modules/m1
2021-03-16T02:52:55.342+0530	error	opa/engine.go:250	error loading rego files{policy path 15 0 /Users/dev-gaur/.terrascan/pkg/policies/opa/rego/aws <nil>} {error 26 0  lstat /Users/dev-gaur/.terrascan/pkg/policies/opa/rego/aws: no such file or directory}
2021-03-16T02:52:55.342+0530	error	runtime/executor.go:121	failed to initialize OPA policy engine
2021-03-16T02:52:55.342+0530	error	cli/run.go:107	scan run failed{error 26 0  failed to initialize OPA policy engine}
```


configfile.toml looks like this :

```
[policy]
path = "/Users/dev-gaur/.terra1"
#rego_subdir = "/Users/dev-gaur/.terra2/pkg/policies/opa/rego"
```

if you comment out either path or rego_subdir... opa engine will error out while trying to compile the rego files.

There should be a check to warn the user to use the use both of these complimentary fields together properly.

### PR DESCRIPTION
**Pull request by dev-gaur**
fixes #550 . fixes #570, fixes #619

This refactor is mainly regarding how config is handled in terrascan code. These above mentioned bug fixes are a ripple effect of this refactor.

Regarding #619 , a warning will be logged every time the user specify only one of the two policy path params, path & rego_subdir . after the warning, the default value will be applied for the unspecified of the two policy path parameter.

Regarding config, the global scoped configuration will be loaded once and only once, into  singleton object, to maintain a single source of truth. To access the config values, getter functions such as config.GetPolicyBasePath( ) must be used.

PS : make sure you don't store the config values in a package scoped variable like below or inside the init( ) method,

```
package demo

var demoPath = config.GetPolicyBasePath( )

func dpDemo( ) {
    ....
}
```
 that won't work as those values are loaded in compile time whereas the config itself is loaded in runtime. So, that compile time loaded variable, demoPath's, value  would be blank.

### CODE REVIEW

File: pkg/config/global_test.go [hunk 1]
```diff
@@ -18,10 +18,14 @@ package config
 
 import (
 	"testing"
+
+	"github.com/accurics/terrascan/pkg/utils"
 )
 
 func TestLoadGlobalConfig(t *testing.T) {
 	testConfigFile := "./testdata/terrascan-config.toml"
+	absDefaultBasePolicyPath, absDefaultPolicyRepoPath, _ := utils.GetAbsPolicyConfigPaths(defaultBasePolicyPath, defaultPolicyRepoPath)
```
Comment (kanchwala-yusuf) [inline pkg/config/global_test.go:27]: It would be better to handle errors here!

File: pkg/initialize/run.go [hunk 2]
```diff
@@ -29,22 +29,23 @@ import (
...
+func Run(isNonInitCmd bool) error {
 	zap.S().Debug("initializing terrascan")
 
+	zap.S().Debugf("rego subdir path : %s", config.GetPolicyRepoPath())
 	// check if policy paths exist
-	if path, err := os.Stat(policyRepoPath); err == nil && path.IsDir() {
-		if isScanCmd {
+	if path, err := os.Stat(config.GetPolicyRepoPath()); err == nil && path.IsDir() {
+
+		zap.S().Debug("EXISTS AND IS A DIR")
```
Comment (kanchwala-yusuf) [inline pkg/initialize/run.go:45]: Can we please improve this debug message? or remove it?

File: pkg/initialize/run.go [hunk 3]
```diff
@@ -29,22 +29,23 @@ import (
...
+		if isNonInitCmd {
+			zap.S().Debug("IS NON INIT")
```
Comment (kanchwala-yusuf) [inline pkg/initialize/run.go:47]: Can we please improve this debug message? or remove it?

File: pkg/notifications/notifiers.go [hunk 4]
```diff
@@ -25,15 +25,11 @@ import (
...
-	notificationsConfigKey = "notifications"
-)
-
 var (
 	errNotifierNotSupported   = fmt.Errorf("notifier not supported")
 	errNotifierTypeNotPresent = fmt.Errorf("notifier type not present in toml config")
-	// ErrTomlKeyNotPresent will be returned when config file does not have notificationsConfigKey
-	ErrTomlKeyNotPresent = fmt.Errorf("key not present in toml config")
+	// ErrNotificationNotPresent error is caused when there isn't any notification present in the config
+	ErrNotificationNotPresent = fmt.Errorf("no notification present")
```
Comment (kanchwala-yusuf) [inline pkg/notifications/notifiers.go:32]: Let's try to indicate that the notifications are not present in config file

File: pkg/notifications/notifiers.go [hunk 5]
```diff
@@ -51,25 +47,15 @@ func NewNotifier(notifierType string) (notifier Notifier, err error) {
...
-	}
-
 	// get config for 'notifications'
-	notifications := configReader.GetNotifications()
-	if notifications == nil {
-		zap.S().Debug("key '%s' not present in toml config", notificationsConfigKey)
-		return notifiers, ErrTomlKeyNotPresent
+	notifications := config.GetNotifications()
+	if len(notifications) == 0 {
+		zap.S().Debug("no notification detected from config")
```
Comment (kanchwala-yusuf) [inline pkg/notifications/notifiers.go:57]: May be we can improve this message, something like `no notification specified in config`

File: pkg/runtime/executor_test.go [hunk 6]
```diff
@@ -18,16 +18,16 @@ package runtime
 
 import (
 	"fmt"
+	iacProvider "github.com/accurics/terrascan/pkg/iac-providers"
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor_test.go:21]: Please run `goimports -e -w pkg/`!

File: pkg/runtime/rules.go [hunk 7]
```diff
@@ -2,39 +2,26 @@ package runtime
...
-	if len(configReader.GetCategory().List) > 0 {
-		e.categories = configReader.GetCategory().List
-	}
+	e.categories = config.GetCategoryList()
 
 	// specify severity of violations to be reported
-	if len(configReader.GetSeverity().Level) > 0 {
-		e.severity = configReader.GetSeverity().Level
+	if len(config.GetSeverityLevel()) > 0 {
+		e.severity = config.GetSeverityLevel()
```
Comment (kanchwala-yusuf) [inline pkg/runtime/rules.go:24]: Please check why `rules_test.go` is deleted?

**Review by kanchwala-yusuf — APPROVED** [review]

File: pkg/initialize/run.go [hunk 8]
```diff
@@ -64,6 +65,12 @@ func Run(isScanCmd bool) error {
 
 // DownloadPolicies clones the policies to a local folder
 func DownloadPolicies() error {
+
+	policyBasePath := config.GetPolicyBasePath()
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:69]: instead of assigning new variables, we can use values directly?

File: pkg/notifications/notifiers.go (hunk 5 above)
Comment (patilpankaj212) [reply pkg/notifications/notifiers.go:57]: `no notifications specified in the config file`

File: pkg/cli/register.go [hunk 9]
```diff
@@ -43,8 +46,24 @@ func Execute() {
...
+		var configfile string
+		if len(ConfigFile) > 0 {
+			configfile = ConfigFile
+		}
+
+		if len(configfile) == 0 {
+			configfile = os.Getenv(configEnvvarName)
+		}
+
+		zap.S().Debugf("%s=%s", configEnvvarName, os.Getenv(configEnvvarName))
```
Comment (patilpankaj212) [inline pkg/cli/register.go:59]: if env variable is not specified, we would log:
`TERRASCAN_CONFIG=`, is that okay?
also, we are using `:` when logging a value.

File: pkg/initialize/run.go [hunk 10]
```diff
@@ -29,22 +29,23 @@ import (
...
 )
 
 const terrascanReadmeURL string = "https://raw.githubusercontent.com/accurics/terrascan/master/README.md"
 
 // Run initializes terrascan if not done already
-func Run(isScanCmd bool) error {
+func Run(isNonInitCmd bool) error {
 	zap.S().Debug("initializing terrascan")
 
+	zap.S().Debugf("rego subdir path : %s", config.GetPolicyRepoPath())
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:41]: with this log, we would be logging rego subdir two times

File: test/e2e/init/init_test.go [hunk 11]
```diff
@@ -30,15 +25,24 @@ import (
...
+	initCommand            = "init"
+	defaultPolicyRepoPath  = filepath.Join(utils.GetHomeDir(), ".terrascan")
+	terrascanGitURL        = "https://github.com/accurics/terrascan.git"
+	terrascanDefaultBranch = "master"
+	terrascanConfigEnvName = "TERRASCAN_CONFIG"
+	kaiMoneyGitURL         = "https://github.com/accurics/KaiMonkey.git"
+
+	testPolicyRepoPath = filepath.Join(utils.GetHomeDir(), ".terrascan-test")
+	testRegoSubDirPath = filepath.Join(testPolicyRepoPath, "pkg", "policies", "opa", "rego")
+	warnNoBasePath     = "policy rego_subdir specified in configfile 'config/relative_rego_subdir.toml', but base path not specified. applying default base path value"
```
Comment (patilpankaj212) [inline test/e2e/init/init_test.go:44]: the configfile in the warn messages should be created using filepath.Join()

**Review by patilpankaj212 — CHANGES_REQUESTED** [review]
added review.

File: pkg/runtime/rules.go (hunk 7 above)
Comment (dev-gaur) [reply pkg/runtime/rules.go:24]: I recall that rules_test.go was checking how rules are being parsed from the config file. and now that logic has been shifted inside LoadGlobalConfig itself now.

File: pkg/initialize/run.go (hunk 10 above)
Comment (dev-gaur) [reply pkg/initialize/run.go]: removed

File: pkg/cli/register.go (hunk 9 above)
Comment (dev-gaur) [reply pkg/cli/register.go]: changes made

File: pkg/initialize/run.go (hunk 8 above)
Comment (dev-gaur) [reply pkg/initialize/run.go:64]: changes made where feasible

File: pkg/runtime/executor_test.go (hunk 6 above)
Comment (dev-gaur) [reply pkg/runtime/executor_test.go:24]: thanks for pointing out

**Review by patilpankaj212 — APPROVED** [review]

**Pull request merged**